src/compiler/compiler.rs

```diff
@@ -894,28 +922,93 @@ where
         f_ok(None)
     };
 
-    let creator = creator.clone();
+
+    let creator1 = creator.clone();
+    let creator2 = creator.clone();
     let executable = executable.to_owned();
-    let env = env.to_owned();
+    let executable2 = executable.clone();
+    let env1 = env.to_owned();
+    let env2 = env.to_owned();
+    let env3 = env.to_owned();
     let pool = pool.clone();
-    Box::new(rustc_vv.and_then(move |rustc_vv| match rustc_vv {
-        Some(Ok(rustc_verbose_version)) => {
-            debug!("Found rustc");
-            Box::new(
-                Rust::new(
-                    creator,
-                    executable,
-                    &env,
-                    &rustc_verbose_version,
-                    dist_archive,
-                    pool,
+    let cwd = cwd.to_owned().clone();
+    Box::new(
+        rustc_vv
+            .and_then(move |rustc_vv| match rustc_vv {
+            Some(Ok(rustc_verbose_version)) => {
+                debug!("Found rustc");
+
+                Box::new(
+                    RustupProxy::find_proxy_executable::<T>("rustup", creator, &env1)
+                        .and_then(move |proxy : Result<RustupProxy>| -> SFuture<(Option<RustupProxy>, PathBuf)> {
```

Comment:
How would you determine if `rustc` is a proxy? Currently it's attempted if `rustup` is installed, and if, if try to see if there is an overwrite of the default toolchain and then move from there.

Reply:
This is what we do in Firefox to figure out whether `rustc` is a proxy: https://searchfox.org/mozilla-central/rev/c79c0d65a183d9d38676855f455a5c6a7f7dadd3/build/moz.configure/rust.configure#23-80

It would be somewhat more complicated than what's already being done, but I think it would be more correctly generally.

Reply:
That would add another layer of indirection, and an additional process to spawn, but I agree that this would be the best approach.
Can we merge this and make that check a subsequent PR?

Reply:
I suggested a simple test in the issue:
https://github.com/mozilla/sccache/issues/87#issuecomment-587547770

Reply:
But it already uses the which crate? That would only resolve if rustup would be right next beside it, but not for the case where rustc is installed via os packages but also a rustc version is installed via rustup. In that case the test as posted in #87 is not yielding viable new information. Hence I would prefer to do this properly as @froydnj suggested, but as a follow up / once everything else in this PR is merge ready.

Reply:
The point is not to detect "is rustup installed at all?" but "is this *particular* rustc we're being invoked with a rustup shim?"

Reply:
Ok, so the assumption here is that `rustup` and `rustc` being in the same dir are a sufficient discriminator that the rust is a shim. While this is cheap I would still prefer the behaviour check.

Reply:
Implemented.